Only allow messages to RemoteGraphicsContextGL when WebGL is enabled

rniwa opened the pull request:
#### bd4fdb4399408b01fc0d96303e4dc466d3c1ca7e
<pre>
Only allow messages to RemoteGraphicsContextGL when WebGL is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=277534">https://bugs.webkit.org/show_bug.cgi?id=277534</a>

Reviewed by Kimmo Kinnunen.

Enable RemoteGraphicsContextGL only when WebGL is enabled and GPU process is used for WebGL.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::RemoteGraphicsContextGL):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
(WebKit::RemoteGraphicsContextGL::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Tools/Scripts/generate-gpup-webgl:

Canonical link: <a href="https://commits.webkit.org/281964@main">https://commits.webkit.org/281964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4506faefbb5cd41f77373506445b210b47f77c40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61596 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14179 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65566 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12137 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63714 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48630 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12410 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49700 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8428 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64664 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38037 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/53356 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30534 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34716 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10579 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/11090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56504 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10884 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67295 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5535 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10640 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57081 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5561 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57299 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4552 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9271 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36748 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37833 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/38927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37578 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
```diff
@@ -24,6 +24,7 @@
 // This file is generated by generate-gpup-webgl. Do not edit.
 #if ENABLE(GPU_PROCESS) && ENABLE(WEBGL)
 
+[EnabledBy=WebGLEnabled,UseGPUProcessForWebGLEnabled]
```
There's a comment at the top of this file that says:
```
// This file is generated by generate-gpup-webgl. Do not edit.
```

This change needs to be made at https://github.com/WebKit/WebKit/blob/f6d5526e6123330465ed88dc52d6325f09e62c1a/Tools/Scripts/generate-gpup-webgl#L89

rniwa replied:
Oh, right, fixing...

rniwa replied:
Fixed.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -106,8 +106,8 @@ RemoteGraphicsContextGL::~RemoteGraphicsContextGL()
     assertIsMainRunLoop();
-    workQueue().dispatch([attributes = WTFMove(attributes), protectedThis = Ref { *this }]() mutable {
-        protectedThis->workQueueInitialize(WTFMove(attributes));
+    workQueue().dispatch([attributes = WTFMove(attributes), sharedPreferencesForWebProcess = m_gpuConnectionToWebProcess.get()->sharedPreferencesForWebProcess(), protectedThis = Ref { *this }]() mutable {
```
I don't think this is neccessary. The `RemoteGraphicsContextGL` is only ever constructed when `WebGLEnabled && UseGPUProcessForWebGLEnabled`.  The real shared preferences object cannot change for the duration of the instance, and the copy won't change also because it's a copy.

This is equivalent of having just "const bool isEnabled = true" and checking that for each of the RGCGL message, hundreds and thousands of message invocations, and having it just return true all the time.

There's nothing that would run the RGCGL message code in case attacker would invoke the message when the setting is off. You already added the critical check is in 

```
messages -> GPUConnectionToWebProcess WantsDispatchMessage {
    ....
   
 [EnabledBy=WebGLEnabled,UseGPUProcessForWebGLEnabled] void CreateGraphicsContextGL(WebKit::GraphicsContextGLIdentifier identifier, struct WebCore::GraphicsContextGLAttributes attributes, WebKit::RenderingBackendIdentifier renderingBackendIdentifier, IPC::StreamServerConnectionHandle serverConnection) AllowedWhenWaitingForSyncReply
    [EnabledBy=WebGLEnabled,UseGPUProcessForWebGLEnabled] void ReleaseGraphicsContextGL(WebKit::GraphicsContextGLIdentifier identifier) AllowedWhenWaitingForSyncReply

```

kkinnunen-apple reviewed (commented):
I don't think this good. Checking the flag for each message is redundant, as the message endpoint, RemoteGraphicsContextGL is the only thing handling the message, and that is created only when the flag holds.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -106,8 +106,8 @@ RemoteGraphicsContextGL::~RemoteGraphicsContextGL()
     assertIsMainRunLoop();
-    workQueue().dispatch([attributes = WTFMove(attributes), protectedThis = Ref { *this }]() mutable {
-        protectedThis->workQueueInitialize(WTFMove(attributes));
+    workQueue().dispatch([attributes = WTFMove(attributes), sharedPreferencesForWebProcess = m_gpuConnectionToWebProcess.get()->sharedPreferencesForWebProcess(), protectedThis = Ref { *this }]() mutable {
```
This function is called during object construction.
```
   : ...
    m_sharedPreferencesForWebProcess(gpuConnectionToWebProcess->sharedPreferencesForWebProcess())
```

The settings could be just read at the constructor, no need to read them after constructor, pass them in work queue callback and assign only at the work queue initialization.

rniwa replied:
Oh, that's much cleaner! Will fix.

kkinnunen-apple reviewed (approved):
I'm agreeing that it's good to annotate the interface as it is annotated in this patch.

I'm not agreeing that the instance-based access control would not be as foolproof as checking each message redundantly. I think the system should just check the added annotation during the message receiver construction time in this case.

I'm not sure how much opinion matters in this matter, so marking as approved.

The preferences instance assignment should be done at constructor time, no need to pass that to the work queue and assign there.

rniwa committed:
Only allow messages to RemoteGraphicsContextGL when WebGL is enabled
https://bugs.webkit.org/show_bug.cgi?id=277534

Reviewed by Kimmo Kinnunen.

Enable RemoteGraphicsContextGL only when WebGL is enabled and GPU process is used for WebGL.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::RemoteGraphicsContextGL):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
(WebKit::RemoteGraphicsContextGL::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Tools/Scripts/generate-gpup-webgl:

Canonical link: https://commits.webkit.org/281964@main
